File: client_outbound/schemas/label_schema.ts
/**
 * Label Schema - Human-in-the-Loop Label Data Contract
 *
 * SCHEMA ONLY - NO LOGIC ALLOWED
 *
 * This interface defines the structure of human corrections/labels sent FROM
 * the client TO the Command Centre for model training.
 *
 * CRITICAL ARCHITECTURE:
 * ----------------------
 * CLIENT ➜ COMMAND CENTRE (ONE-WAY OUTBOUND)
 *
 * - Client sends human labels/corrections to Command Centre
 * - Labels are aggregated by Command Centre for training
 * - Client does NOT learn from labels locally
 * - Client does NOT receive model updates through this interface
 * - No model updates occur on the client
 * - All training happens in the Command Centre
 *
 * This is a pure data contract with no validation, logic, or runtime behavior.
 *
 * DO NOT:
 * - Add validation logic
 * - Add runtime behavior
 * - Add local learning logic
 * - Add model distribution logic
 * - Implement training on the client
 */

/**
 * Label status values.
 */
export type LabelStatus = "correct" | "changed" | "unsure";

/**
 * Human-in-the-loop label payload sent from client to Command Centre.
 *
 * Represents a human correction or confirmation of a detection event.
 * These labels are sent to the Command Centre where they are aggregated
 * for future model training. The client never performs training locally.
 */
export interface HumanLabelSchema {
  /**
   * ID of the detection event being labeled.
   * References a previously sent DetectionEventSchema.
   */
  event_id: string;

  /**
   * The corrected license plate text as provided by the human operator.
   * Null if the human determined no plate was present.
   */
  corrected_plate: string | null;

  /**
   * Label status indicating the type of human action taken.
   *
   * - "correct": Human confirmed the detection was correct
   * - "changed": Human corrected the detection
   * - "unsure": Human was unsure and flagged for review
   */
  label_status: LabelStatus;

  /**
   * When the human provided this label (ISO 8601 format).
   * Example: "2024-01-15T14:35:00Z"
   */
  labeled_at: string;

  /**
   * Optional metadata about the labeling action.
   */
  metadata?: {
    /**
     * ID of the user who provided the label.
     */
    user_id?: string;

    /**
     * Time taken by the human to provide the label (in seconds).
     */
    labeling_duration_seconds?: number;

    /**
     * Optional notes from the human operator.
     */
    notes?: string;

    /**
     * Original detected text for reference.
     */
    original_plate_text?: string | null;

    /**
     * Original confidence score for reference.
     */
    original_confidence?: number;

    /**
     * Additional context about the labeling action.
     */
    [key: string]: any;
  };
}

/**
 * REMINDER: This is a schema-only file.
 *
 * - No validation logic
 * - No runtime behavior
 * - No learning or training on client
 * - Labels are sent to Command Centre only
 * - Client does not learn from labels
 * - No model updates occur on client
 * - All training happens in Command Centre
 */

File: command_centre/schemas/label_schema.py
"""
Label Schema - Human-in-the-Loop Label Data Contract

SCHEMA ONLY - NO LOGIC ALLOWED

This module defines the structure of human corrections/labels received BY the
Command Centre FROM client deployments.

CRITICAL ARCHITECTURE:
----------------------
CLIENT ➜ COMMAND CENTRE (ONE-WAY INBOUND)

- Command Centre receives human labels/corrections from clients
- Labels are aggregated for future model training
- No learning occurs here yet (this is schema only)
- This defines expected inbound label payloads from clients
- Model distribution will be PUSH-ONLY in the future (separate channel)
- Clients never receive model updates through this interface

This is a pure data contract with no validation, logic, or runtime behavior.

DO NOT:
- Add validation logic
- Add runtime behavior
- Import pydantic, dataclasses, or validation libraries
- Implement model training here
- Add model distribution logic
- Add bi-directional communication

NOTE: Do NOT import pydantic or other validation libraries.
This is a pure schema definition only.
"""

from typing import Optional, Dict, Any, Literal


# Type alias for label status values
LabelStatus = Literal["correct", "changed", "unsure"]


class HumanLabelSchema:
    """
    Human-in-the-loop label payload received from client deployments.

    SCHEMA ONLY - NO VALIDATION OR LOGIC

    Represents a human correction or confirmation of a detection event.
    These labels are received from clients and aggregated by the Command
    Centre for future model training. No training occurs on the client.

    Attributes:
        event_id: ID of the detection event being labeled
        corrected_plate: The corrected license plate text (None if no plate)
        label_status: Type of human action taken
        labeled_at: When the human provided this label (ISO 8601 format)
        metadata: Optional metadata about the labeling action
    """

    # Type annotations only - NO implementation
    event_id: str  # References a previously sent detection event
    corrected_plate: Optional[str]  # None if no plate present
    label_status: LabelStatus  # "correct" | "changed" | "unsure"
    labeled_at: str  # ISO 8601 format: "2024-01-15T14:35:00Z"
    metadata: Optional[Dict[str, Any]]  # Optional additional context


class HumanLabelMetadata:
    """
    Optional metadata for human labels.

    SCHEMA ONLY - NO VALIDATION OR LOGIC

    Additional context about the labeling action that may be useful for
    quality analysis and training.

    Attributes:
        user_id: ID of the user who provided the label
        labeling_duration_seconds: Time taken to provide the label
        notes: Optional notes from the human operator
        original_plate_text: Original detected text for reference
        original_confidence: Original confidence score for reference
    """

    # Type annotations only - NO implementation
    user_id: Optional[str]
    labeling_duration_seconds: Optional[float]
    notes: Optional[str]
    original_plate_text: Optional[str]
    original_confidence: Optional[float]


"""
REMINDER: This is a schema-only file.

- No validation logic
- No runtime behavior
- No learning or training implementation
- Defines expected inbound label payloads from clients
- Model distribution will be PUSH-ONLY in future (separate channel)
- Clients do not learn from labels
- No model updates sent to clients through this interface
- All training happens in Command Centre only
- This file only documents the data contract
"""
